resolwe/flow/models/fields.py

```diff
@@ -0,0 +1,158 @@
+        placeholder which is safe to insert into SQL query and second
+        one may include potentially dangerous values and must be passed
+        to SQL query in ``params`` attribute to make sure it is properly
+        escaped.
+        """
+        constraints_expression = []
+        constraints_values = []
+        for other_field_name in self.unique_with:
+            if '__' in other_field_name:
+                raise NotImplementedError(
+                    '`unique_with` constraint does not support lookups by related models.'
+                )
+
+            other_field = instance._meta.get_field(other_field_name)  # pylint: disable=protected-access
+            other_value = getattr(instance, other_field_name)
+
+            # Convert value to the database representation.
+            other_field_db_value = other_field.get_prep_value(other_value)
+
+            constraints_expression.append("{} = %s".format(other_field.column))
+            constraints_values.append(other_field_db_value)
+
+        if not constraints_expression:
+            return '', []
+
+        constraints_expression = 'AND ' + ' AND '.join(constraints_expression)
+        return constraints_expression, constraints_values
+
+    def _get_populate_from_value(self, instance):
+        """Get the value from ``populate_from`` attribute."""
+        if hasattr(self.populate_from, '__call__'):
+            # ResolweSlugField(populate_from=lambda instance: ...)
+            return self.populate_from(instance)
+        else:
+            # ResolweSlugField(populate_from='foo')
+            attr = getattr(instance, self.populate_from)
+            return attr() if callable(attr) else attr
+
+    def pre_save(self, instance, add):
+        """Ensure slug uniqunes before save."""
+        slug = self.value_from_object(instance)
+
+        if not slug and self.populate_from:
+            slug = self._get_populate_from_value(instance)
+
+        if slug:
+            slug = slugify(slug)
+
+        if not slug:
+            slug = None
+
+            if not self.blank:
+                slug = instance._meta.model_name  # pylint: disable=protected-access
+            elif not self.null:
+                slug = ''
+
+        if slug:
+            # Make sure that potential sequence added won't excede maximal length.
+            slug = slug[:(self.max_length - 10)]
+
+            constraints_placeholder, constraints_values = self._get_unique_constraints(instance)
+
+            instance_pk_name = instance._meta.pk.name  # pylint: disable=protected-access
+
+            query_params = {
+                'constraints_placeholder': constraints_placeholder,
+                'slug_column': self.column,
+                'table_name': self.model._meta.db_table,  # pylint: disable=protected-access
+                'pk_neq_placeholder': 'AND {} != %s'.format(instance_pk_name) if instance.pk else ''
+            }
+
+            query_escape_params = [slug]
```

Comment:
You could instead use named parameters so this definition would be less fragile.